### ISSUE TITLE
Documentation about accessing AWS Cognito on Android WITHOUT Amplify wrapper/NPM/Compose

### ISSUE THREAD

**Issue by Orbyt** [aws-amplify/amplify-android#2826]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-android/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-android/discussions).

### Language and Async Model

Kotlin or Java

### Amplify Categories

Authentication

### Gradle script dependencies

Not applicable.

### Environment information

Not applicable.

### Please include any relevant guides or documentation you're referencing

AWS Cognito SDK docs (https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-integrate-apps.html), Amplify docs, and more.

### Describe the feature request

I have spent many hours reading documentation, searching Google, online forums, previous issues, and more, before posting here. Note that the "discussions" link for this project linked in the issues template (https://github.com/aws-amplify/amplify-android/discussions) returns a 404.

Documentation about using AWS Cognito _without_ Amplify seem to be missing. The [official AWS Cognito documentation](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-integrate-apps.html) almost exclusively references AWS Amplify. The part of that document referring to AWS Cognito SDKs for Java links to [https://sdk.amazonaws.com/java/.../CognitoIdentityProviderClient](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/cognitoidentityprovider/CognitoIdentityProviderClient.html), which then links to generic SDKs for AWS, none of which seem to be for accessing AWS Cognito in Android (the Java SDK appears to be for Java server applications).

Adding NPM/Typescript/Compose to access Cognito as required by the [AWS Amplify for Android documentation](https://docs.amplify.aws/android/start/quickstart/) is not viable for the project which prompted this question, and considering the multiple StackOverflow and Reddit threads on this topic this seems to be an issue for others as well. Is there any other way to simply authenticate a user in Android with AWS Cognito _without_ making the entire application an "Amplify app"? Answers and documentation for such use-cases would be extremely helpful.

I am more than happy to add additional details to this query if needed. Thank you!

**Comment by mattcreaser** [comment]
Hi @Orbyt thanks for posting your question!

There seems to be a bit of confusion here, I hope I can clarify. The short answer is that yes, it is definitely possible! The most straightforward way to access Cognito without using Amplify is to directly use the underlying [AWS SDK for Kotlin](https://aws.amazon.com/sdk-for-kotlin/) - this is what Amplify itself uses to implement its Cognito functionality. There are some specific examples of using the Kotlin SDK to invoke Cognito APIs available [here](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/kotlin_cognito-identity-provider_code_examples.html). The [Kotlin SDK GitHub repository](https://github.com/awslabs/aws-sdk-kotlin) may be a good starting point for learning more.

Note that the Kotlin SDK is a lower-level library that is specifically useful for invoking AWS APIs, and Amplify builds on top of this to offer higher-level functionality such as storing credentials for you, or automatically refreshing tokens when required. If you use the Kotlin SDK directly you may need to implement some of this functionality yourself.

Moving on to Amplify, none of NPM/Typescript/Compose are strictly **required** to use Amplify to access Cognito. Here's what those technologies are used for:

- `npm`/`typescript`: These are used in Amplify Gen2 to define your backend. Gen2 is built on [Amazon CDK](https://docs.aws.amazon.com/cdk/v2/guide/home.html) which is a type of _infrastructure as code_, so you define the AWS resources you need in TypeScript and then the Gen2 CLI can deploy everything in the cloud. It also outputs a configuration file that you add to your Android application (`amplify_outputs.json`), which is what Amplify Android uses to connect to those resources. The only thing that goes into your application is the configuration file, TypeScript/NPM are only used to define the backend and is not part of your app. Although this is certainly the recommended way to set up your backend, you can still use the [Gen1 CLI tool to setup Auth](https://docs.amplify.aws/gen1/android/build-a-backend/auth/set-up-auth/) instead.
- `compose`: Compose is only used by [amplify-ui-android](https://github.com/aws-amplify/amplify-ui-android/), which are UI components that you can use to build apps using Amplify (in the case of Cognito, it's the Authenticator component). Using Authenticator is strictly optional - you can absolutely choose to build your own UI instead and call through to Amplify Auth APIs to handle the Authentication. This is well covered in the [Amplify Auth docs](https://docs.amplify.aws/android/build-a-backend/auth/connect-your-frontend/sign-in/).

Hope that answers your question, but let me know if something needs further clarification!

**Comment by BigBen3** [comment]
Hi don't know if this will answer your question but I found a repo with cognito and android and pretty sure its without amplify: https://github.com/nameisjayant/Aws-Cognito-User-Pool-Authnetication

**Comment by mattcreaser** [comment]
> Hi don't know if this will answer your question but I found a repo with cognito and android and pretty sure its without amplify: https://github.com/nameisjayant/Aws-Cognito-User-Pool-Authnetication

This app is using the [AWS SDK for Android](https://github.com/aws-amplify/aws-sdk-android), which was the underlying AWS SDK for Amplify v1.x. Although this SDK can still be used, we recommend you either use Amplify (v2.x) or directly access the Kotlin SDK instead.

**Comment by Orbyt** [comment]
> Hi @Orbyt thanks for posting your question!
> 
> There seems to be a bit of confusion here, I hope I can clarify. The short answer is that yes, it is definitely possible! The most straightforward way to access Cognito without using Amplify is to directly use the underlying [AWS SDK for Kotlin](https://aws.amazon.com/sdk-for-kotlin/) - this is what Amplify itself uses to implement its Cognito functionality. There are some specific examples of using the Kotlin SDK to invoke Cognito APIs available [here](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/kotlin_cognito-identity-provider_code_examples.html). The [Kotlin SDK GitHub repository](https://github.com/awslabs/aws-sdk-kotlin) may be a good starting point for learning more.
> 
> Note that the Kotlin SDK is a lower-level library that is specifically useful for invoking AWS APIs, and Amplify builds on top of this to offer higher-level functionality such as storing credentials for you, or automatically refreshing tokens when required. If you use the Kotlin SDK directly you may need to implement some of this functionality yourself.
> 
> Moving on to Amplify, none of NPM/Typescript/Compose are strictly **required** to use Amplify to access Cognito. Here's what those technologies are used for:
> 
>     * `npm`/`typescript`: These are used in Amplify Gen2 to define your backend. Gen2 is built on [Amazon CDK](https://docs.aws.amazon.com/cdk/v2/guide/home.html) which is a type of _infrastructure as code_, so you define the AWS resources you need in TypeScript and then the Gen2 CLI can deploy everything in the cloud. It also outputs a configuration file that you add to your Android application (`amplify_outputs.json`), which is what Amplify Android uses to connect to those resources. The only thing that goes into your application is the configuration file, TypeScript/NPM are only used to define the backend and is not part of your app. Although this is certainly the recommended way to set up your backend, you can still use the [Gen1 CLI tool to setup Auth](https://docs.amplify.aws/gen1/android/build-a-backend/auth/set-up-auth/) instead.
> 
>     * `compose`: Compose is only used by [amplify-ui-android](https://github.com/aws-amplify/amplify-ui-android/), which are UI components that you can use to build apps using Amplify (in the case of Cognito, it's the Authenticator component). Using Authenticator is strictly optional - you can absolutely choose to build your own UI instead and call through to Amplify Auth APIs to handle the Authentication. This is well covered in the [Amplify Auth docs](https://docs.amplify.aws/android/build-a-backend/auth/connect-your-frontend/sign-in/).
> 
> 
> Hope that answers your question, but let me know if something needs further clarification!

@mattcreaser I cannot thank you enough for your well thought out response. I am still going through some of the documentation you linked, so I apologize for the delayed reply. You mentioned an `amplify_outputs.json` file, which is also referenced in the official documentation, but I have not yet found an example of what that file actually looks like (I've also searched this repository) or how to configure it.

For context, the project that prompted this question is using AWS Cognito for user authentication. The React frontend for the project's associated web app is using the `<Authenticator />` component, however I do not believe that the project is using an "Amplify backend". There is no `amplify_outputs.json` file in the web project, and [the button used to download that file as referenced in the documentation ](https://docs.amplify.aws/images/gen2/getting-started/react/nextImageExportOptimizer/amplify-outputs-download-opt-1920.WEBP) does not seem to be present.

Alternatively, there appears to be an `Amplify.configure` function in the SDK, but I am unsure how to use without the aforementioned file. I did find some sort of schema for the `amplify_outputs.json` [in the documentation](https://docs.amplify.aws/android/reference/amplify_outputs/).

In summary, could you explain how to either go about obtaining an `amplify_outputs.json` file when the project is not using an "Amplify backend", or how to configure the Amplify SDK in code?

**Comment by BigBen3** [comment]
> K for Amplify v1.x. Although this SDK can still be used, we recommend you either use Amplify (v2.x) or directly access the Kotlin SDK instead.

Is there any fulll project examples with mvvm that use Kotlin SDK instead?

**Comment by mattcreaser** [comment]
> Is there any fulll project examples with mvvm that use Kotlin SDK instead?

I'm not aware of any such examples. The Kotlin SDK has [a basic Android sample app](https://github.com/scmacdon/aws-doc-sdk-examples/tree/main/kotlin/usecases/subpub_app_android) but it's not MVVM. In general however I would recommend implementing your application as per clean architecture guidelines and use the Kotlin SDK at the repository layer if you wanted to go that route.

> In summary, could you explain how to either go about obtaining an amplify_outputs.json file when the project is not using an "Amplify backend", or how to configure the Amplify SDK in code?

To connect to existing resources without setting up any Amplify cloud resources it's probably easier to use the older `amplifyconfiguration.json` file format. You can copy/paste the JSON from the [documentation page here](https://docs.amplify.aws/gen1/android/build-a-backend/auth/existing-resources/) and fill in your Cognito information.

You can omit the `OAuth` object if you're only doing basic SRP auth. 

To configure Amplify with this JSON data you have a couple of options:
- Place the JSON content in a file at `<module>/src/res/raw/amplifyconfiguration.json`. Then you can call `Amplify.configure(context)` and it will automatically read this file.
- Instantiate an instance of `AmplifyConfiguration` and pass it to the configure function: `Amplify.configure(AmplifyConfiguration.fromJson(JSONObject(jsonString)), context)`.

Hope that helps!

**Comment by mattcreaser** [comment]
There are examples showing how to use the Kotlin SDK to perform various operations with Cognito [here](https://github.com/awsdocs/aws-doc-sdk-examples/tree/main/kotlin/services/cognito#code-examples). It's for running on the JVM, not an Android application, but the code to use the Kotlin SDK is the same.
